specs/lazy-adr/adr-018-forced-inclusion-mechanism.md

```diff
@@ -0,0 +1,462 @@
+    txs         map[common.Hash]DirectTransaction  // Map of direct transactions
+    mu          sync.RWMutex                       // Mutex for thread-safe access
+    latestDAHeight uint64                          // Latest DA block height scanned
+}
+```
+
+#### Sequencer Status Tracking
+
+```go
+type SequencerStatus struct {
+    IsActive          bool      // Whether the sequencer is considered active
+    LastActiveHeight  uint64    // Last DA height where sequencer posted a batch
+    DownSince         time.Time // Time since sequencer was last seen active
+}
+```
+
+### APIs and Interfaces
+
+#### Enhanced DA Client Interface
+
+```go
+type DAClient interface {
+    // Existing methods
+    // ...
+    
+    // New method for forced inclusion
+    GetDirectTransactions(ctx context.Context, fromHeight, toHeight uint64) ([][]byte, error)
+    // Note: SubmitDirectTransaction is removed as it's not a responsibility of the rollup node
+}
+```
+
+#### Sequencer Interface Extensions
+
+```go
+// New methods added to the Sequencer interface
+func (s *Sequencer) ScanDALayerForDirectTxs(ctx context.Context) error
+func (s *Sequencer) IncludeDirectTransactions(ctx context.Context, batch *Batch) error
+func (s *Sequencer) ValidateDAReference(ctx context.Context, header *types.Header) error
+```
+
+#### Full Node Interface Extensions
+
+```go
+// New methods added to the Node interface
+func (n *Node) CheckSequencerStatus(ctx context.Context) (bool, error)
+func (n *Node) ProcessDirectTransactions(ctx context.Context) error
+func (n *Node) ValidateBlockWithDAReference(ctx context.Context, block *types.Block) error
+```
+
+### Implementation Changes
+
+#### Single Sequencer Node Changes
+
+1. **DA Layer Scanner**:
+   - Implement a periodic scanner that queries the DA layer for direct transactions
+   - Track all direct transactions in the DirectTxTracker data structure
+   - Update the latest scanned DA height after each scan
+
+2. **Transaction Inclusion Logic**:
+   - Modify the batch creation process to include direct transactions from the DA layer
+   - Ensure all direct transactions from a DA block are included before moving to the next DA block
+   - Add the DA block reference information (height, hash, timestamp) to the `ExtraData` field in the rollup block header
+
+3. **Validation Rules**:
+   - Implement timestamp validation to ensure the rollup block's timestamp is not significantly higher than the referenced DA block
+   - Prevent re-referencing of the same DA block across multiple rollup blocks
+
+4. **Recovery Mechanism**:
+   - Add logic to detect when the sequencer comes back online after downtime
+   - Implement state synchronization to catch up with any forced inclusions that occurred during downtime
+   - Resume normal operation by building on top of the canonical chain tip
+
+#### Sequencer Operation Flow
+
+The following diagram illustrates the operation flow for the sequencer with forced inclusion:
+
+```
```

Comment:
can we migrate these diagram to mermaid, its easier to maintain longterm. Even though they may not look as nice